File: problems/missing-number/missing-number.cs
using System;
using System.Linq;

namespace ConsoleApplication
{
   public class MissingNumberExample
    {
        public static void Main()
        {
            MissingNumber(new int[] { -3,-2,-1,0,1,3 }); 

            Console.ReadKey();
        }
 
      public static void MissingNumber(int[] arr)
        {
            int sum = arr.Sum();
            arr = arr.OrderBy(i => i).ToArray();
           int incremental_sum = ((arr.Length + 1) * (arr[arr.Length - 1] + arr[0])) / 2;
            Console.WriteLine($"The missing number is: { incremental_sum - sum}");
        }
    }
}

File: problems/missing-number/missing-number.rs
//returns the missing number, or in the case Vec is empty, returns None
fn missing_number(input: &[usize]) -> Option<usize> {
    if input.is_empty() {
        return None;
    }

    let non_missing: usize = (1..input.len() + 2).sum();
    let missing: usize = input.iter().sum();

    Some(non_missing - missing)
}

fn main() {
    println!(
        "missing number: {}",
        missing_number(&[1, 2, 3, 4, 6]).unwrap()
    );
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn returns_none_if_vec_is_empty() {
        assert_eq!(None, missing_number(&[]));
    }

    #[test]
    fn finds_missing_number_sequential_range() {
        assert_eq!(4, missing_number(&[1, 2, 3, 5, 6]).unwrap());
    }
}
